src/encoded/static/components/motifs.js

```diff
@@ -0,0 +1,158 @@
+        throw new Error('not ok');
+    }).catch((e) => {
+        console.log('OBJECT LOAD ERROR: %s', e);
+    });
+}
+
+// Display information on page as JSON formatted data
+export class MotifElement extends React.Component {
+    constructor() {
+        super();
+
+        this.addMotifElement = this.addMotifElement.bind(this);
+    }
+
+    componentDidMount() {
+        require.ensure(['d3', 'd3-sequence-logo'], (require) => {
+            this.d3 = require('d3');
+            this.sequenceLogos = require('d3-sequence-logo'); // logos This is for local development when changes are needed to d3-sequence-logo.
+
+            const targetElement = this.chartdisplay;
+            const entryPoint = this.sequenceLogos.entryPoint;
+
+            getMotifData(this.pwmLink, this.context.fetch).then((response) => {
+                this.addMotifElement(targetElement, response, entryPoint);
+            });
+        });
+    }
+
+    // Redraw charts when window changes
+    componentDidUpdate() {
+        this.d3 = require('d3');
+        this.sequenceLogos = require('d3-sequence-logo'); // logos This is for local development when changes are needed to d3-sequence-logo.
+
+        const targetElement = this.chartdisplay;
+        const entryPoint = this.sequenceLogos.entryPoint;
+
+        // Need to remove current logo or another logo will be appended to original
+        this.chartdisplay.innerHTML = '';
+
+        getMotifData(this.pwmLink, this.context.fetch).then((response) => {
+            this.addMotifElement(targetElement, response, entryPoint);
+        });
+    }
+
+    addMotifElement(targetElement, response, entryPoint) {
+        // Convert PWM text data to object
+        const PWM = convertTextToObj(response);
+        // Generate the logo from the PWM object
+        entryPoint(targetElement, PWM, this.d3);
+    }
+
+    render() {
+        const element = this.props.element;
+        const targetList = element.targets.map(t => t.label);
+        const targetListLabel = targetList.length > 1 ? 'Targets' : 'Target';
+        const pwmLink = `${this.props.urlBase}${element.documents[0]['@id']}${element.documents[0].attachment.href}`;
+
+        return (
+            <div className="element" id={`element${element.accession}`}>
+                <div className="motif-description">
+                    <p><a href={element['@id']}>{element.accession}</a></p>
+                    {element.biosample_term_name ?
+                        <p>Biosample: {element.biosample_term_name}</p>
+                    : null}
+                    {element.organ_slims ?
+                        <p>Organ: {element.organ_slims.join(', ')}</p>
+                    : null}
+                    {targetList ?
+                        <p>{targetListLabel}: {targetList.join(', ')}</p>
+                    : null}
+                </div>
+                <div ref={(div) => { this.chartdisplay = div; this.pwmLink = pwmLink; }} className="motif-element" />
```

Comment:
The `this.pwmLink` assignment in the ref function is very unusual since it’s not related to the ref, and will break in React 16 since ref functions don’t exist anymore. Since you only use this.pwmLink in a couple places, I think you might as well just calculate `pwmLink` from scratch whenever you need it, or even put the string template into a one-line method so you only need to change it once if it changes.

Reply:
…or you could put the assignment in the constructor if you know urlBase and documents won’t change between page re-renders. I don’t know if they can change or not.

Reply:
I moved this and I think it should work better.